spec_util/meld.go

```diff
@@ -614,3 +580,34 @@ func (m *melder) meldList(dst, src *pb.List) error {
 	}
 	return nil
 }
+
+// Assumes dst.value == src.value.
+// Meld data formats, tracking data, etc. from src to dst.
+// XXX(cns): In some cases, this modifies src as well as dst :/
+func (m *melder) meldPrimitive(dst, src *pb.Primitive) {
+	// Special case: If and only if one data has a type hint, assign it to the other
+	// data so that the difference does not trigger a conflict and the type hint is preserved.
+	// XXX(cns): This modifies src!  Not ideal, but I don't know if it's safe
+	// to remove this behavior.
+	{
+		if src.TypeHint != dst.TypeHint {
+			if dst.TypeHint == "" {
+				dst.TypeHint = src.TypeHint
+			} else if src.TypeHint == "" {
+				src.TypeHint = dst.TypeHint
+			}
+		}
+	}
+
+	// Merge data formats
+	mergedDataFormats := make(map[string]bool, len(src.Formats)+len(dst.Formats))
+	for k := range src.Formats {
+		mergedDataFormats[k] = true
+	}
+	for k := range dst.Formats {
+		mergedDataFormats[k] = true
+	}
+	if len(mergedDataFormats) > 0 {
+		dst.Formats = mergedDataFormats
+	}
+}
```

Comment:
When merging a formatted int with an unformatted int, the result should be an unformatted int. This appears to produces a formatted int. Similarly for other types, of course.

Reply:
As it stands, primitives are only merged if the base type and format kind are equal; the cases you're talking about currently result in a type conflict.